<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Week 14 - Putting it all together &amp; next steps</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
    <link rel="stylesheet" href="slides_copy.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center





# *Week 14 &amp;ndash; &lt;br&gt; Putting it all together &amp; next steps*

----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra
### 2021/04/13 (updated: 2021-04-12)

---
class:inverse middle center

# I: Putting it all together

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## What did we cover?

.pull-left[
- **Shell**
    - Basics
    - Data tools
    - Scripting
]

.pull-right[
- **Python**
    - Basics
    - Coding approaches
    - Regular expressions
    - NumPy and Pandas
    - Biopython
]

- Version control with **Git**

- Submitting OSC jobs using **SLURM**

- Reproducible workflows using **Snakemake**

- **Project organization and management**

- Formatted plain text files with **Markdown**

- Along the way, you have gotten to know the **VS Code** editor quite well

---

## What is the big picture?

Core goal of the course, as described in the first week:

&gt; **Learning skills to do your research more reproducibly and efficiently**

---

## What is the big picture?

Apply these **habits and skills**:

- Use one directory per project with a clear and consistent sub-dir structure
  and file naming.

- Treat raw data as read-only.

- Document everything you do in plain text README files.

--

- Do as much as possible using code:
  - Including what you could in some cases do as/more quickly using GUIs &amp;ndash;
    downloads, renaming and re-organizing, systematic literature searches, etc.:
    "Use code as documentation."
  
  - Use shell tools for "simple" problems (one or a few steps),  
    use Python as it gets more complex. 

- Always use version control.

- Share your code (GitHub).

---

## What is the big picture? (cont.)

This course has focused on **foundational skills** and
overall on **breadth over depth**.

Some of the rationale for this:

- Specific tools are not only highly ... specific, they are also easier and perhaps
  better picked up in short "just-in-time workshops" or even using good cookbook-style
  tutorials (increasingly common!).
  Moreover:
  
  &gt; Learning core bioinformatics data skills will give you the
  &gt; foundation to learn, apply, and assess any bioinformatics program or analysis
  &gt; method. In 10 years, bioinformaticians may only be using a few of the bioinformatics
  &gt; software programs around today. But we most certainly will be using data skills and
  &gt; experimentation to assess data and methods of the future.

--

- The first steps are the hardest &amp;ndash; and "false starts" can be very common
  when trying to self-teach tools that will initially take more time than
  manual approaches (e.g. version control, keep renaming and doing other
  repetitive things by hand).

- You may not be aware of what's out there.

---

## What is the big picture? (cont.)

Flipside:

- Limited practical case studies

- You may not feel you've become really proficient with any of these tools

&lt;br&gt;

&gt; *It is important to practice regularly, and protect time to play with new tools.
&gt; While it is great to have a specific task in mind when starting to program,
&gt; it is also important to leave time to acquire and improve new skills.*

&gt; *You will be happier if you think “I have invested two hours in learning Python!”
&gt; rather than “It took me two hours to solve this easy problem in Python—
&gt; I could have done it in Excel in 20 minutes!” While this might be true when
&gt; you start, your investment will pay back in the long run.*

&gt; *With a bit of perseverance, you will start reusing code to automate processes,
&gt;  thereby saving a lot of time.
&gt; Even better, you will tackle problems that you could not have addressed at all
&gt; without your new skills.* &amp;mdash; CSB Chapter 11

---

## Some misc. tips going forward

- Avoid the tar pit of immediacy.

- Break things up in small bits.

- Write code to be read by humans and data to be read by computers.

- Develop frequently used scripts into "tools",
  which you can can keep in a separate repository.

- Make figures, tables, and statistics the result of scripts.

- Use existing libraries whenever possible.

- Keep practicing!

---
class:inverse middle center

# II: Next steps

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Some specific next steps by topic &amp;ndash; Python

- Shell -- we covered quite bit, focus on getting fluent.  
  Some things we didn't cover:
  
  - [Creating links](https://mcic-osu.github.io/pracs-sp21/posts/week-02/slides/02-3-shfiles.html#44)
  - `find` and `xargs` ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-05/slides/05-3_find-xargs.html#1) &amp;ndash; see Buffalo chapter 7)
  - [`join`](https://mcic-osu.github.io/pracs-sp21/posts/week-04/slides/04-1_data-tools.html#55) to merge tabular data by a shared column
  - Parameter expansion
  - Process management (background jobs)
  - tar archives
  - If you like shell scripts, look into parsing arguments
  - bioawk
  
---

## Some specific next steps by topic &amp;ndash; Python

- Python &amp;ndash; If you want to keep using Python, I would probably recommend
  working your way through a book fully focused on Python.
  
  See the resources page -- Type of book depends on your preference/direction:
  general Python vs Data Science vs Bioinformatics
  
  Jupyter Notebooks

- Git

  - Merge conflicts
  - Undoing things ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-03/slides/03-3_git.html#1))

- Use VS Code locally and SSH(-tunnel) in

---

## Further resources

As you have hopefully seen, there already was a page with lots of further 
resources for Python.

That page now also has further resources for all other topics.

...add link...

---

## Other foundational skills

If you want to keep doing "data-centric" work,
be that in genomics / bioinformatics or data science,
I would also recommend a couple of topics we did not touch on at all:

- Statistics

- Data visualization

- R &amp;ndash; certainly for bioinformatics and probably also the best choice for
  the two topics above.

---

## OSU Courses
  
- [Microbiome Science Training Track](https://u.osu.edu/coms/for-trainees/microbiome-informatics-training-track/)
  by the Center of Microbiome Science

- [Biomedical Informatics](https://medicine.osu.edu/departments/biomedical-informatics/education/bmi-courses) courses

- [STAT 5730: Introduction to R for Data Science](https://stat.osu.edu/courses/stat-5730)

- [PLNTPTH 7003.01: Agricultural Genomics](https://plantpath.osu.edu/courses/plntpth-700301)

- HCS7004: Genome Analytics

- HCS7806: Data Visualization (Jessica Cooperstone)

.content-box-info[
For an overview of "data science and analytics" courses at OSU,  
see [this TDAI page](https://tdai.osu.edu/programs/).
]

.content-box-info[
Also see:
- [Microbiome Informatics webinar series](https://u.osu.edu/coms/webinar-series/)
  with hands-on materials by the Center of Microbiome Science.

- [Code Club](https://biodash.github.io/codeclub/) &amp;ndash;
  weekly meeting, hopefully for credits this fall.
]

---

## Workshops

**Shorter workshops (mostly 2-3 days):**

- [Bioinformatics.ca](https://bioinformaticsdotca.github.io/)
  &amp;ndash; Yearly workshops on topics like microbiome analysis, RNAseq, etc.

- [Physalia Courses](https://www.physalia-courses.org/courses-workshops/)

- The Carpentries &amp;ndash; e.g. the
  [Genomics Curriculum of Data Carpentry](https://datacarpentry.org/lessons/#genomics-workshop)  
  (Since earlier this year there is also an [OSU chapter](https://tdai.osu.edu/carpentries-at-osu/).)

&lt;br&gt;

**Longer workshops (&gt; 1 week):**

- [Summer Institute in Statistical Genetics](https://si.biostat.washington.edu/suminst/)

- [Evomics](http://evomics.org/): "Molecular Evolution", "Genomics",
  "Phylogenomics", etc.

- [CSH Programming for Biologists](http://programmingforbiology.org/)

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
