<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Week 14 - Putting it all together &amp; next steps</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <script src="libs/fabric/fabric.min.js"></script>
    <link href="libs/xaringanExtra-scribble/scribble.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-scribble/scribble.js"></script>
    <script>document.addEventListener('DOMContentLoaded', function() { window.xeScribble = new Scribble({"pen_color":["#FF0000"],"pen_size":3,"eraser_size":30}) })</script>
    <link href="libs/panelset/panelset.css" rel="stylesheet" />
    <script src="libs/panelset/panelset.js"></script>
    <script src="libs/clipboard/clipboard.min.js"></script>
    <link href="libs/xaringanExtra-clipboard/xaringanExtra-clipboard.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-clipboard/xaringanExtra-clipboard.js"></script>
    <script>window.xaringanExtraClipboard(null, {"button":"Copy Code","success":"Copied!","error":"Press Ctrl+C to Copy"})</script>
    <script src="libs/mark.js/mark.min.js"></script>
    <link href="libs/xaringanExtra-search/search.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-search/search.js"></script>
    <script>window.addEventListener('load', function() { window.xeSearch = new RemarkSearch({"position":"bottom-left","caseSensitive":false,"showIcon":true,"autoSearch":true}) })</script>
    <link rel="stylesheet" href="slides.css" type="text/css" />
    <link rel="stylesheet" href="slides_copy.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center





# *Week 14 &amp;ndash; &lt;br&gt; Putting it all together &amp; next steps*

----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra
### 2021/04/13

---
class:inverse middle center

# I: Putting it all together

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## What did we cover?

.pull-left[
- **Shell**
    - Basics
    - Data tools
    - Scripting
    - OSC jobs with SLURM
]

.pull-right[
- **Python**
    - Basics
    - Coding approaches
    - Regular expressions
    - NumPy and Pandas
    - Biopython
]

--

- **Project organization and management**

- Formatted plain text files with **Markdown**

- Version control with **Git**

- Reproducible workflows using **Snakemake**

- Along the way, you have gotten to know the **VS Code** editor quite well

---

## What is the big picture?

Core goal of the course:

.content-box-green[
**Learning strategies and computational skills to do your research
more reproducibly and efficiently**
]

---

## Some take-home messages

Apply these **strategies and skills**:

- Use one directory per project with a clear and consistent sub-dir structure
  and ditto file naming.

- Treat raw data as read-only.

- Document everything you do in plain text README files.

- Always use version control.

- Share your code (using GitHub).

---

## Some take-home messages

- Do as much as possible using **code**:

  - Use code even for tasks you could in some cases do as/more quickly using GUIs &amp;ndash;
    downloads, renaming and re-organizing, systematic literature searches, etc.
    
  - "Use code as documentation."
  
  - Use shell tools for "simple" problems (consisting of one or a few steps),  
    use Python as tasks gets more complex. 

- Break your code up in small, reusable bits:  
  keep individual scripts short, use functions in Python, etc.

- Write code to be read by humans, and data to be read by computers.

---

## Foundational skills

This course has focused on *foundational skills* and *breadth over depth*.  

Some of the rationale for this &amp;ndash; **foundational skills**:

&gt; *Learning core bioinformatics data skills will give you the*
&gt; *foundation to learn, apply, and assess any bioinformatics program or analysis*
&gt; *method. In 10 years, bioinformaticians may only be using a few of the bioinformatics*
&gt; *software programs around today. But we most certainly will be using data skills and*
&gt; *experimentation to assess data and methods of the future.*  
&gt; &amp;mdash; Buffalo chapter 1

--

- High turn-over of specific methods...

- With foundational skills&lt;sup&gt;[1]&lt;/sup&gt;, more specialized methods/tools can be
  picked up with "just-in-time" workshops or even using good online tutorials.
  
- These skills are broadly applicable, and useful in any data-heavy occupation.

.footnote[
&lt;sup&gt;[1]&lt;/sup&gt; Though not just the one ones taught here, as we'll get back to.
]

---

## Breadth over depth

This course has focused on *foundational skills* and *breadth over depth*.

Some of the rationale for this &amp;ndash; **breadth over depth**:

- The first steps are the hardest &amp;ndash; and "false starts" can be very common
  when trying to self-teach tools that will initially take more time than
  manual approaches.
  
  E.g. it is tempting to keep pushing off using version control,  
  or to keep renaming and doing other repetitive things by hand.

- You may not be aware of what's out there.

&lt;br&gt;

--

But there is a flip side to this:

- We've seen few practical case studies.

- You may not feel you've become really proficient with any of these tools.

---
class:inverse middle center

# II: Next steps

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Various tips going forward

- Keep practicing!

- Avoid the "tar pit of immediacy".

&gt; *It is important to practice regularly, and protect time to play with new tools.
&gt; While it is great to have a specific task in mind when starting to program,
&gt; it is also important to leave time to acquire and improve new skills.*

&gt; *You will be happier if you think “I have invested two hours in learning Python!”
&gt; rather than “It took me two hours to solve this easy problem in Python—
&gt; I could have done it in Excel in 20 minutes!” While this might be true when
&gt; you start, your investment will pay back in the long run.*

&gt; *With a bit of perseverance, you will start reusing code to automate processes,
&gt;  thereby saving a lot of time.
&gt; Even better, you will tackle problems that you could not have addressed at all
&gt; without your new skills.* &amp;mdash; CSB Chapter 11

---

## Various tips going forward (cont.)

- Develop frequently used scripts into "tools",
  which you can can keep in a separate repository.

- Don't try to re-invent the wheel but use existing libraries / packages / programs
  for specialized tasks (unless it's just for practice).
  
- **Try to make all figures, tables, and statistics in papers to be
  the result of scripts.**

---

## Some specific next steps by topic

.content-box-info[
As you have hopefully seen, there already was a page on the coure's GitHub site
with further resources for Python.

I have now changed [that page](https://mcic-osu.github.io/pracs-sp21/resources.html)
to hold further resources for all other topics.
]

---

## Some specific next steps by topic &amp;ndash; Shell

- We actually covered quite a bit here!
  
  While you could go a lot deeper, for things
  well beyond the complexity that we discussed,
  other tools may be better choices:
  - Python (or R) &amp;ndash; for more complex data tasks 
  - Snakemake &amp;ndash; for more complex workflow tasks
  
  Instead, I would recommend to focus on getting fluent and picking new things
  up along the way.

---

## Some specific next steps by topic &amp;ndash; Shell

- But because we had pretty extensive coverage on the shell,
  it may make sense to provide a list of commonly-used shell tools and topics
  we didn't cover:
  
  - A command-line text editor like *Nano* or *Vim*.
  - Creating symbolic links
    ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-02/slides/02-3-shfiles.html#44))
  - `find` and `xargs`
    ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-05/slides/05-3_find-xargs.html#1)
    &amp;ndash; and see Buffalo chapter 7)
  - `join` to merge tabular data 
    ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-04/slides/04-1_data-tools.html#55)
    &amp;ndash; and see Buffalo chapter 7)
  - `bioawk` (for sequence data &amp;ndash; see Buffalo chapter 7)
  - Manual software installation and `$PATH`
  - Process management and monitoring (Buffalo chapter 4)
  - tar archives
  - Parameter expansion
  - If you like shell scripts, look into parsing arguments with `getopts`
  
---

## Some specific next steps by topic (cont.)

### Python
  
- If you want to keep using Python, I would probably recommend
  working your way through a **book fully focused on Python.**  
    
  See the resources page &amp;ndash; the choice of book will depends on your
  preferred direction: general Python vs Bioinformatics vs Data Science.
  
- Jupyter Notebooks

&lt;br&gt;

--

### Git

- Merge conflicts
  ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-03/slides/03-2_git.html#60)).
  
- Undoing things
  ([slides](https://mcic-osu.github.io/pracs-sp21/posts/week-03/slides/03-3_git.html#1)).

---

## Some specific next steps by topic (cont.)

### Miscellaneous
  
- Use **VS Code locally** and SSH(-tunnel) in.
  
- Learn about using Docker/Singularity containers.

- If you're using Windows,
  consider switching (or dual-booting with) Ubuntu Linux.
  
---

## Other foundational skills

If you want to keep doing "data-centric" work,
be that in genomics / bioinformatics or data science,
I would also recommend digging into a couple of topics that we did not touch on at all:

- Statistics

- Data visualization

- R &amp;ndash; certainly for bioinformatics and probably also the best choice for
  the two topics above.

&lt;br&gt;

.content-box-info[
Reading the rest of the two books will get you started with the latter two!
]

---

## OSU Courses
  
- [Microbiome Science Training Track](https://u.osu.edu/coms/for-trainees/microbiome-informatics-training-track/)
  by the Center of Microbiome Science

- [Biomedical Informatics](https://medicine.osu.edu/departments/biomedical-informatics/education/bmi-courses) courses

- [STAT 5730: Introduction to R for Data Science](https://stat.osu.edu/courses/stat-5730)

- [PLNTPTH 7003.01: Agricultural Genomics](https://plantpath.osu.edu/courses/plntpth-700301)

- HCS7004: Genome Analytics (Jonathan Fresnedo-Ramirez)

- HCS7806: Data Visualization (Jessica Cooperstone)

&lt;br&gt;

.content-box-info[
For an overview of "data science and analytics" courses at OSU,  
see [this TDAI page](https://tdai.osu.edu/programs/).
]

---

## Workshops

**Shorter workshops (mostly 2-3 days):**

- [Bioinformatics.ca](https://bioinformaticsdotca.github.io/)
  &amp;ndash; Yearly workshops on topics like microbiome analysis, RNAseq, etc.

- [Physalia Courses](https://www.physalia-courses.org/courses-workshops/)

- The Carpentries &amp;ndash; e.g. the
  [Genomics Curriculum of Data Carpentry](https://datacarpentry.org/lessons/#genomics-workshop)  
  (Since earlier this year there is also an [OSU chapter](https://tdai.osu.edu/carpentries-at-osu/).)

&lt;br&gt;

**Longer workshops (&gt; 1 week):**

- [Summer Institute in Statistical Genetics](https://si.biostat.washington.edu/suminst/)

- [Evomics](http://evomics.org/): "Molecular Evolution", "Genomics",
  "Phylogenomics", etc.

- [CSH Programming for Biologists](http://programmingforbiology.org/)

---

## Misc. at OSC

.content-box-info[
Also see:
- [Microbiome Informatics webinar series](https://u.osu.edu/coms/webinar-series/)
  with hands-on materials by the Center of Microbiome Science.

- [Code Club](https://biodash.github.io/codeclub/) &amp;ndash;
  weekly meeting, hopefully for credits this fall.
]

---

## Recaps / bonus

- [SSH set-up and tunneling](https://mcic-osu.github.io/pracs-sp21/w06_UA_ssh.html)

- [Using Conda with Snakemake](https://mcic-osu.github.io/pracs-sp21/posts/week-13/slides/13-3.html#conda)

- JupyterLab at OSC

---
class: inverse middle center

# Questions?

----
&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
