<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Week 10 - Writing Good Code</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/fontawesome-all.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
    <link rel="stylesheet" href="slides_copy.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center



# *Week 10 - Writing good code*

----

# I: &lt;br&gt; Modules, functions, &lt;br&gt; program structure, and writing style

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra
### 2021/03/16 (updated: 2021-03-08)

---

## Overview of this week

This presentation:

- Modules

- Functions

- Program structure

- Writing style

---

&lt;br&gt;

Thursday's presentation:

- Error handling

- Debugging

- Unit testing

- Profiling

---
class: center middle inverse

# Modules

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Modules

Modules, packages, and libraries:

- **Python modules** contain *collections of functions*.  
  Technically, Python modules are single `.py` files.
  
- *Collections of modules* are technically **packages**,
  though they can also be simply referred to as modules.   

- The term **library** is not as well-defined,
  and can be used interchangeably with both module and package.


--

&lt;br&gt;

Modules and packages can be *imported* into your Python environment
(in a script or interactively) using the import statement:
  
  ```py
  import os
  ```

---

## Installing modules

Modules from the **Python standard library** don't have to be installed,
and can be imported directly.

So, while they are part of any Python installation, they are not loaded by
default. 

----

--

So-called "**third-party**" modules/packages/libraries, on the other hand,
do need to be installed.

These are not installed from inside Python (like with R),
but using stand-alone package managers, usually in the shell:

- We have already worked with `conda` &amp;ndash; to install the 
  `numpy` ("NUMerical PYthon") library, for instance:
  ```sh
  $ conda install numpy
  ```
  
- Another common package manager is the Python-specific `pip`:
  ```sh
  $ pip install numpy
  ```

---

## Loading modules

In Python, it is customary to not load modules directly into your "namespace".

This means that the module's functions will not be accessibly by just their names,
but you need to prepending the module's name.

- Simply using the module name as in `import mymodule` will load all of the module's
  function, which can then be referred to using `mymodule.myfunction`:

  ```py
  import os
  os.getcwd()
  ```

- If a module's name is long(ish), it is common to define an "alias" for the
  module in the `import` call:
  
  ```py
  import numpy as np
  my_array = np.array([1, 2, 3, 4])
  ```
  
  (While the alias name can be anything you like,
  most modules have commonly-used aliases, like `np` for `numpy`.)  

---

## Loading modules (cont.)

- To only load certain functions from a moudle,
  or certain modules from multi-module packages:

  ```py
  # From the snakemake.io module, import two functions:
  from snakemake.io import expand, glob_wildcards
  
  # From the matplotlib package, import the pyplot module: 
  from matplotlib import pyplot as plt
  
  # Or equivalently using &lt;package&gt;.&lt;module&gt;:
  import matplotlib.pyplot as plt
  ```

&lt;br&gt;

- Finally, it *is* possible but discouraged to load all functions from a module
  directly into your namespace using an asterisk `*` wildcard:
  
  ```sh
  from numpy import *
  array([1, 2, 3, 4])
  ```

---

## Creating modules

In Python, it is exceedingly easy to create a module: just save a file with
functions (e.g. `my_utils.py`) and load it using `import my_utils`  
(assuming the file is in the same directory or can be otherwise found).

---
class: center middle inverse

# Functions

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Writing our first function

```py
def GC_content(dna):

    # Make sure the input string is all uppercase:
    dna = dna.upper()
    
    # Count the occurrences of each nucleotide:
    numG = dna.count("G")
    numC = dna.count("C")
    numA = dna.count("A")
    numT = dna.count("T")
    
    # Finally, calculate (G + C) / (A + T + G + C):
    GC = (numG + numC) / (numG + numC + numT + numA)
    
    return GC 
```

---

## Writing our first function (cont.)

```py
def GC_content(dna):

    # Arbitrary code...
    
    return GC
```

The anatomy of a Python function:

- Start with the keyword `def` (for *define*)

- Next provide the name of the function, which is whatever we would like it to be.
  
- Between parenthesis, list the input(s) (*argument(s)*) of the function.

- Don't forget the colon `:`!

- All the code belonging to the function is indented.

- The function returns output as defined after the `return` keyword.

---

## Using our first function (cont.)

```python
GCcontent("AATTTCCCGGGAAA")
#&gt; 0.42857142857142855

GCcontent("ATGCATGCATGC")
#&gt; 0.5
```

---

## Why are custom functions useful?

Custom functions are useful to break up your code in small(ish) units.

This way, the main part of the program can become a series of logically
connected function calls,
and the nuts and bolts are the individual functions.

In other words, having modular code:

- Makes it easier to see what your program does.

- Makes it easier to test your code.

- Makes it easier to repeat operations
  (like calculating GC content for mulitple sequences).

- Makes your programs more flexible (function arguments!) and
  easier to adapt.

---

## A few more simple functions

A function to nicely print a dictionary:

```python
def print_dictionary(mydic):
    
    for k, v in mydic.items():
        print("key: ", k, " value: ", str(v))
```

&lt;br&gt;

--

Let's try it:

```python
print_dictionary({"a": 3.4, "b": [1, 2, 3, 4], "c": "astring"})

#&gt; key:  a  value:  3.4
#&gt; key:  b  value:  [1, 2, 3, 4]
#&gt; key:  c  value:  astring
```

---

## A few more simple functions (cont.)

A function to calculate the square for a series of numbers,
and put those squares in a list:
  
```python
def squared(start = 1, end = 10):
      
    # Create empty list to catch result of each iteration:
    results = []
      
    # Loop through the numbers:
    for i in range(start, end):
        r = i ** 2        # Calculate square 
        results.append(r) # Append current value to list
      
    return results
```

Let's try it:

```python
squared(start = 3, end = 10)
#&gt; [9, 16, 25, 36, 49, 64, 81]
```

---

## A few more simple functions (cont.)

```python
def squared(start = 1, end = 10):
    ...
```

Because we specified default values in the function definition,  
we can omit one or both!

```python
# Specify only start, end has default value 10:
squared(5)
#&gt; [25, 36, 49, 64, 81]

# Specify only end, start has default value 1:
squared(end = 3)
#&gt; [1, 4]

# Start has default value 1, end is 10:
squared()
#&gt; [1, 4, 9, 16, 25, 36, 49, 64, 81]
```

---

## <i class="fas  fa-user-edit "></i> Intermezzo 4.1: Determine the function

```python
def foo1(x = 7):
    return x ** 0.5
```

--

`foo1()` returns the square root of its argument.

&lt;br&gt;

--

```python
def foo2(x = 3, y = 5):
    if x &gt; y:
        return x
    return y
```

--

`foo2()` returns the larger of two numbers.

---

## <i class="fas  fa-user-edit "></i> Intermezzo 4.1: Determine the function (cont.)


```python
def foo4(x = 6):
    result = 1
    for i in range(1, x + 1):
        result = result * i
    return result
```

--

`foo4()` returns the factorial of a number.

&lt;br&gt;

--

```python
def foo6(x = 25):
    if x == 1:
        return 1
    return x * foo6(x - 1)
```

--

`foo6()` returns the factorial of a number via *recursion*.

---

## <i class="fas  fa-user-edit "></i> A function to convert temperature

- Define a function that converts from temperatures in degrees Fahrenheit to
  degrees celcius.

  (To convert Fahrenheit to Celcius, subtract 32 and then multiply by 5/9.)

- Test your function with any temperature.

---

## <i class="fas  fa-user-edit "></i> A function to convert temperature: Solution

```python
def f_to_c(deg_f):
    deg_c = (deg_f - 32) * (5 / 9)
    return deg_c
```

```python
#&gt; f_to_c(100)
37.77777777777778
```

---
class: center middle inverse

# Program structure

-----

### Practicing coding a more complex program &lt;br&gt; via a population genetic simulation

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;


---

## The outline of the program

We will simulate genetic drift: random changes in allele frequency over time
that happen in finite populations.

We'll ask how long it takes before an allele drifts to fixation.

--

&lt;br&gt;

To do so, we'll divide our code into several functions:

- A function that initializes the population:
  It should take as input the size of the population (`N`),
  and the probability of having an `A` allele (`p`).
  This function returns an entire population.

- A function that computes genotype frequencies,
  which we will need to determine whether an allele has gone to fixation.
  The function should take a population as input and output the count for each
  genotype.

- A reproduction function that takes the current population and produces the
  next generation.

&lt;br&gt;

--

**Start by opening a new file and saving it as `drift.py`.**

---

## The function to build the population

```python
import scipy  # for random numbers

def build_population(N, p):
    """The population consists of N individuals. 
       Each individual has two chromosomes, containing
       allele "A" or "a", with probability p and 1-p, respectively.
       The population is a list of tuples.
    """
    
    population = []                  # Start with empty list
    
    for i in range(N):               # Loop over individuals
        
        allele1 = "A"                # The "default" allele is A
        if scipy.random.rand() &gt; p:  # Random number between 0 and 1
            allele1 = "a"
        
        allele2 = "A"                # Second chromosome (diploidy)
        if scipy.random.rand() &gt; p:
            allele2 = "a"
        
        population.append((allele1, allele2))
    
    return population
```

---

## Test the function (not in the script)

```python
build_population(N = 5, p = 0.5)
#&gt; [('a', 'a'), ('a', 'A'), ('A', 'A'), ('A', 'a'), ('A', 'A')]

build_population(N = 5, p = 0.5)
#&gt; [('A', 'a'), ('A', 'a'), ('a', 'a'), ('A', 'A'), ('A', 'a')]

build_population(N = 5, p = 0.9)
#&gt; [('A', 'A'), ('A', 'A'), ('A', 'a'), ('A', 'A'), ('A', 'A')]

build_population(N = 5, p = 0)
#&gt; [('a', 'a'), ('a', 'a'), ('a', 'a'), ('a', 'a'), ('a', 'a')]
```

---

## The function to calculate genotype frequencies

```python
def compute_frequencies(population):
    """ Count the genotypes.
        Returns a dictionary with counts for each genotype.
    """
    
    AA = population.count(("A", "A"))
    Aa = population.count(("A", "a"))
    aA = population.count(("a", "A"))
    aa = population.count(("a", "a"))
    
    return({"AA": AA,
            "aa": aa,
            "Aa": Aa,
            "aA": aA})
```

---

## Test the function (not in the script)

```python
&gt;&gt;&gt; my_pop = build_population(6, 0.5)
&gt;&gt;&gt; compute_frequencies(my_pop)
```

---

## The function to create a new generation
```python
def reproduce_population(population):
    """ Create a new generation through sexual reproduction
        For each of N new offspring:
        - Choose the parents at random
        - The offspring receives one chromosome from each parent
    """
    
    new_generation = []
    
    N = len(population)
    
    for i in range(N):
        
        dad = scipy.random.randint(N)  # random integer between 0 and N-1
        mom = scipy.random.randint(N)
        
        chr_mom = scipy.random.randint(2)  # which chromosome comes from mom
        
        offspring = (population[mom][chr_mom],
                     population[dad][1 - chr_mom])
        
        new_generation.append(offspring)
    
    return(new_generation)
```

---

## Test the function (not in the script)

```python
&gt;&gt;&gt; reproduce_population(my_pop)
```

---

## A new script for the main program

- Open a new file and save it as `simulate_drift.py`.

- In it, we will first import our functions:

  ```python
  import drift
  ```

- Let's check whether this worked:

  ```python
  &gt;&gt;&gt; help(drift)
  ```

---

```python
def simulate_drift(N, p):
    
    # Initialize the population:
    my_pop = drift.build_population(N, p)
    
    num_generations = 0
    fixation = False
    
    while fixation == False:
    
      # Compute genotype counts:
      genotype_counts = drift.compute_frequencies(my_pop)
      
      # If one allele went to fixation, stop:
      if genotype_counts["AA"] == N or genotype_counts["aa"] == N:
          print("An allele reached fixation at generation", num_generations)
          print("The genotype counts are")
          print(genotype_counts)
          
          fixation == True
          break # Superfluous?
      
      # If not, reproduce:
      my_pop = drift.reproduce_population(my_pop)
      num_generations = num_generations + 1
```

---

## Test the program

```python
simulate_drift(100, 0.5)
#&gt; An allele reached fixation at generation 66
#&gt; The genotype counts are
#&gt; { ' aa ' : 100, ' aA ' : 0, ' AA ' : 0, ' Aa ' : 0}

simulate_drift(100, 0.9)
#&gt; An allele reached fixation at generation 20
#&gt; The genotype counts are
#&gt; { ' aa ' : 0, ' aA ' : 0, ' AA ' : 100, ' Aa ' : 0}
```

---

## Saving and loading Python objects

```python
import pickle
# save Python object to file
pickle.dump(my_pop, open("population.pickle", "wb"))
# What is wb?
```

```python
population = pickle.load(open("population.pickle", "rb"))
print(population)
```

---
class: center middle inverse

# Writing style 

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Some style  guidelines

LINK TO STYLE GUIDELINES

- Use four spaces per indentation level; not tabs.

- Split long lines (especially the arguments of functions) using parentheses.

- Separate functions using blank lines.

- Use a new line for each operation:

.pull-left[
Yes:
```python
import scipy
import re
```
]

.pull-left[
No:
```python
import scipy,re
```
]

---

## Some style guidelines (cont.)

- Put spaces around operators (unless using the : to “slice” a list); use
parentheses to clarify the precedence of operators:

.pull-left[
Yes:
```python
a = b ** c
z = [a, b, c, d]
n = n * (n - 1)
k = (a * b) + (c * d)
my_list = a_list[1:n]
```
]

.pull-right[
No:
```python
a=b**c
z=[ a,b,c,d ]
n =n*(n-1)
k=a*b+c*d
my_list = a_list[1 : n]
```
]

---

## Some style guidelines (cont.)

- **Use lots of comments and "docstrings"!**

  Python allows you to write short comments (starting with # ) and
  docstrings (multiline comments flanked by triple double quotes, """ ):
  
  - Use docstrings to document *how to use the code*.
    What does this function do? What is the input/output?
  
  - Use short comments for explaining difficult passages in the code,
    that is, to document how the code works.


&lt;br&gt;

- **Naming conventions:**

  - Variables and functions in lowercase, "snake case": `my_var`, `my_fun`.
  
  - **Constants** in uppercase: `SOME_CONSTANT`.
  
  - Use descriptive names: `count_pairs`, `body_mass`, `cell_volume`.
  
  - Verbs for functions and nouns for objects.

---
class: center middle inverse

# Questions?

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
