<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Week 11 - Python: Scientific computing</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
    <link rel="stylesheet" href="slides_copy.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center





# *Week 11 - Python: Scientific computing*

----

# III: BioPython

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra
### 2021/03/25 (updated: 2021-03-19)


---
class: center middle inverse

# BioPython (CSB 6.4)

-----

# I: Retrieving sequences from NCBI (6.4.1)

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Using BioPython to inferface with NCBI databases

While we already downloaded some data from NCBI manually in the exercises for
week 7 (Bioinformatics Data),
we can also use NCBI's **"Entrez Programming Utilities"** to do this
programmatically.

BioPython includes an Entrez module that interfaces with NCBI Entrez. 
This is a much better method to find and download data, if you regularly need
to do so.

- We start by importing the module:
  
  ```python
  from Bio import Entrez
  ```

- As another setup step,
  we should provide our e-mail address to let NCBI know who we are:

  ```python
  Entrez.email = "me.1@osu.edu"  # Replace by your email address
  ```

---

## Search the database

- Create a "handle" to search results:

  ```python
  handle = Entrez.esearch(db = "nuccore",
                          term = ("Uropsilus investigator[Organism]"),
                          RetMax = 200)
  ```

- Read the results and close the handle:

  ```python
  search_results = Entrez.read(handle)
  handle.close()
  ```

&lt;br&gt;

- Our results are returned in the form of a dictionary,
  so let's look at the keys:
  
  ```python
  search_results.keys()
  #&gt; dict_keys(['Count', 'RetMax', 'RetStart', 'IdList',
  #&gt;            'TranslationSet', 'TranslationStack',
  #&gt;             'QueryTranslation'])
  ```
  
---

## Looking at the results

- How many sequences did we find?  
  
  ```python
  search_results["Count"]
  '126'   # Has increased from '71' in the book!
  ```
  
- What is the list of GenBank identifiers?

  ```python
  id_list = search_results["IdList"]
  print(id_list)
  #&gt; ['524853022', '555947199', '555947198', ... , '555946814']
  ```

- Did we get all 126 IDs in the list? Note that the default is to only return
  20, but because we used `RetMax = 200` in the call to `Entrez.esearch()`,
  we got all of them:

  ```python
  len(id_list)
  #&gt; '126'
  search_results["RetMax"]
  #&gt; '200'
  ```

---

## Downloading the sequences

We'll now use the GenBank IDs to actually download some sequences:

```python
search_handle = Entrez.efetch(db = "nuccore",
                              rettype = "fasta",
                              retmode = "text",
                              id = id_list)

# Set up a handle to an output file:
out_handle = open("Uropsilus_seq.fasta", "w")

# Write the obtained sequence data to the output file:
for line in handle:
    out_handle.write(line)

# Close the handles:
out_handle.close()
search_handle.close()
```

Next, we'll read in our FASTA file using SeqIO.

---

class: center middle inverse

# BioPython (CSB 6.4)

-----

# II: Input and output of sequence data using *SeqIO* (6.4.2)

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## Reading a FASTA file with SeqIO

- We start by importing the SeqIO module:

  ```python
  from Bio import SeqIO
  ```

- Let's print the ID and sequence length for each sequence:

  ```python
  in_handle = open("Uropsilus_seq.fasta", "r")
  
  for record in SeqIO.parse(in_handle, "fasta"):
      print(record.description)
      print(len(record), '\n')
  in_handle.close()
  
  #&gt; KC516837.1 Uropsilus investigator isolate A11 apolipoprotein B (ApoB) gene, partial cds
  #&gt; 573 
  
  #&gt; KC516819.1 Uropsilus investigator voucher mlxs331 cytochrome c oxidase subunit I (COI) gene, partial cds, alternatively spliced; mitochondrial
  #&gt; 912 
  
  #&gt; KC516818.1 Uropsilus investigator voucher mlxs022 cytochrome c oxidase subunit I (COI) gene, partial cds, alternatively spliced; mitochondrial
  #&gt; 912 
  #&gt; [...]
  ```

---

## Selecting only sequences for the BMI1 genes

```python
# We will import the "re" module for regular expression searches:
import re

output_handle = open("Uropsilus_BMI1.fasta", "w")

for record in SeqIO.parse("Uropsilus_seq.fasta", "fasta"):
    
    # Find the BMI1 sequences using "re.search"
    if re.search("BMI1", record.description):
        print(record.id)
        short_seq = record[:100]  # Just take the first 100 bases
        SeqIO.write(short_seq, output_handle, "fasta")
        
output_handle.close()
```

---

class: center middle inverse

# BioPython (CSB 6.4)

-----

# III: Programmatic BLAST Search (6.4.3)

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

## BLAST

The **Basic Local Alignment Search Tool (BLAST)** finds regions of similarity
between biological sequences,
and is one of the most widely used bioinformatics tool.

You may not it from its web interface, where you can paste in or upload some
sequences, and then "BLAST them" against a database you can specify.

### screenshot?


&lt;br&gt;

Using this web interface may be convenient when you have one or a few sequences,
but it is not very reproducible and completely unsuitable for automation.

---

## Getting set up for BLAST

Biopython provides the module `Bio.Blast` which we can use to run a BLAST search
against the NCBI's online databases.

We will import just a few of its functions: `NCBIWWW` for access to NCBI's
BLAST server (the BLAST analysis is still being performed there!),
and `NCBIXML` to parse the XML format that our results will be in

```python
# from Bio import SeqIO   # Should already be loaded
# from Bio import Entrez  # Should already be loaded
from Bio.Blast import NCBIWWW, NCBIXML
```

&lt;br&gt;

--

We will start by parsing our FASTA file with SeqIO,
and then converting the records into a list:

```python
handle = open("Uropsilus_BMI1.fasta", "r")
records = list(SeqIO.parse(handle, "fasta"))
```

---

## Our sequence records as a named list

What does this list that we created look like? Let's print the first 2 items:

```python
records[0:2]
#&gt; [SeqRecord(seq=Seq('TATTATGCTGTTTTGTGAACCTGTAGAAAACAAGTGCTTTTTATCTTGAAATTC...CCA', SingleLetterAlphabet()), id='KF778086.1', name='KF778086.1', description="KF778086.1 Uropsilus investigator voucher KIZ:020539 polycomb ring finger oncoprotein (BMI1) gene, 3' UTR", dbxrefs=[]),
#&gt; SeqRecord(seq=Seq('TATTATGCTGTTTTGTGAACCTGTAGAAAACAAGTGCTTTTTATCTTGAAATTC...CCA', SingleLetterAlphabet()), id='KF778085.1', name='KF778085.1', description="KF778085.1 Uropsilus investigator voucher KIZ:020527 polycomb ring finger oncoprotein (BMI1) gene, 3' UTR", dbxrefs=[])]
```

### More about this list type

Say that the fourth sequence is the one we are interested in:

```python
print(records[3].id, " ", records[3].seq)
#&gt; KF778083.1   TATTATGCTGTTTTGTGAACCTGTAGAAAACAAGTGCTTTTTATCTTGAAATTCAACAAATGGAAAGAATATGCATAGAATAATGCATTCTATGTAGCCA
```

---

## Run the BLAST search

For our BLAST search, we will use the nucleotide database (`nt`)
with the `blastn` algorithm, one of several BLAST variants
(other options are `blastp`, `blastx`, `tblastn`, and `tblastx`).

The `qblast()` function requires three arguments: the program (`blastn` for us),
the database (`nt` for us), and the query sequence (our 4th record from the
FASTA file): 

```python
Entrez.email = "me.999@osu.edu"

result_handle = NCBIWWW.qblast("blastn", "nt", records[3].seq)
```

Now, we will write our results, which are in XML format, to an XML file
and then close the result handle:

```python
save_file = open("my_blast.xml", "w")
save_file.write(result_handle.read())
save_file.close()

result_handle.close()
```

---

## Parse the XML file

We will now parse our XML file with blast results using the `NCBIXML` parser:

```python
result_handle = open("my_blast.xml")
blast_records = NCBIXML.read(result_handle)
```

(Because we our query consisted only of a single sequence,
we could use `NCBIXML.read()`.
For multiple sequences, you should use `NCBIXML.parse`.)

---

## Processing the BLAST results

Next, we loop through the individual alignment hits and then for each hit
through the "High-Scoring Pairs" (HSPs, subsequences with a good match).

For each High-Scoring Pair, we check whether the match is good enough
according to an arbitrary E-value (`E_VALUE_THRESH`) and length (`LEN_THRESH`)
that we set.

```python
E_VALUE_THRESH = 0.04
LEN_THRESH = 3000

for alignment in blast_records.alignments:
    
    for hsp in alignment.hsps:
        
        if hsp.expect &lt; E_VALUE_THRESH and alignment.length &gt; LEN_THRESH:
            
            print("****Alignment****")
            print("sequence:", alignment.title)
            print("length:", alignment.length)
            print("E value:", hsp.expect)
            print(hsp.query[0:75] + "...")
            print(hsp.match[0:75] + "...")
            print(hsp.sbjct[0:75] + "...")
```

---

## Processing the BLAST results

```python
E_VALUE_THRESH = 0.04
LEN_THRESH = 3000

for alignment in blast_records.alignments:
    
    for hsp in alignment.hsps:
        
        if hsp.expect &lt; E_VALUE_THRESH and alignment.length &gt; LEN_THRESH:
            
            print("****Alignment****")
            print("sequence:", alignment.title)
            print("length:", alignment.length)
            print("E value:", hsp.expect)
            print(hsp.query[0:75] + "...")
            print(hsp.match[0:75] + "...")
            print(hsp.sbjct[0:75] + "...")
#&gt; ****Alignment****
#&gt; sequence: gi|1304911126|ref|XM_006933246.4| PREDICTED: Felis catus BMI1 proto-oncogene, polycomb ring finger (BMI1), transcript variant X3, mRNA
#&gt; length: 3523
#&gt; E value: 2.25861e-42
#&gt; TATTATGCTGTTTTGTGAACCTGTAGAAAACAAGTGCTTTTTATC...
#&gt; |||||||||||||||||||||||||||||||||||||||||||||...
#&gt; TATTATGCTGTTTTGTGAACCTGTAGAAAACAAGTGCTTTTTATC...
#&gt; [...]
```

---
class: center middle inverse

# Questions?

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;


---
class: center middle inverse

# Bonus material

-----

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---

class: center middle inverse

# BioPython (CSB 6.4)

-----

# IV: Querying PubMed for scientific literature information (6.4.4)

&lt;br&gt; &lt;br&gt; &lt;br&gt; &lt;br&gt;

---
background-color: #f2f5eb

## PubMed search

Using the `Entrez.esearch()` function we saw earlier,
we can search any NCBI database,
including PubMed, which contains scientific literature.

By way of example, we will search for any papers on *Drosophila* that mention
the gene "*spaetzle*" anywhere in the title or abstract:

```python
from Bio import Entrez
Entrez.email = "me.999@osu.edu"

handle = Entrez.esearch(db = "pubmed",
                        term = ("spaetzle[Title/Abstract] AND Drosophila[ALL]"),
                        usehistory = "y")
record = Entrez.read(handle)
handle.close()                        
```                        

--

- We used `usehistory = "y"`, which will allows us to refer back to our search
  to fetch the titles and abstracts by saving `WebEnv` and `QueryKey`:

  ```python
  webenv = record["WebEnv"]
  query_key = record["QueryKey"]
  ```

---
background-color: #f2f5eb

## PubMed search (cont.)

- How many hits did we get?

  ```python
  record["Count"]
  #&gt; '15'
  ```
  
  We found 15 records (up from 13 in the book) that contained the words
  "*spaetzle*" and "*Drosophila*".
  
- We can now fetch the titles and abstracts:

  ```python
  handle = Entrez.efetch(db = "pubmed",
                           rettype = "medline", retmode = "text",
                           webenv = webenv, query_key = query_key)
  data = handle.read()
  handle.close() 
  ```

- Finally, we write the results to file:

  ```python
  out_handle = open("Spaetzle_abstracts.txt", "w")
  out_handle.write(data)
  out_handle.close()
  ```

---
background-color: #f2f5eb

## PubMed search (cont.)

- Let's have a look at the results:

  ```python
  !cat Spaetzle_abstracts.txt
  ```
  
- With a simply `grep` command, we can just see lines
  (and some context with `-C 1`) with the word "*Spaetzle*":  

  ```python
  !grep -i "spaetzle" -C 1 Spaetzle_abstracts.txt
  
  #&gt;     leading to ventrally-restricted expression of the sulfotransferase Pipe. These
  #&gt;    events promote the ventral processing of Spaetzle, a ligand for Toll, which
  #&gt;    ultimately sets up the embryonic dorsal-ventral axis. We then describe the
  #&gt; --
  #&gt; DP  - 2019 Nov 12
  #&gt; TI  - Dynamics of Spaetzle morphogen shuttling in the Drosophila embryo shapes
  #&gt;    gastrulation patterning.
  #&gt; --
  #&gt;    The dynamics indicate that a sharp extracellular gradient is formed through
  #&gt;    diffusion-based shuttling of the Spaetzle (Spz) morphogen that progresses through
  #&gt;    several nuclear divisions. Perturbed shuttling in wntD mutant embryos results in 
  ```

---
background-color: #f2f5eb

## PubMed search (cont.)

Alternatively, we could use regular expressions in Python to nicely get each *sentence* with "*Spaetzle*" along with the PubMedID for the publication:

```python
import re

with open("Spaetzle_abstracts.txt") as datafile:
    pubmed_input = datafile.read()
    # Delete newlines followed by 6 white spaces
    # ... to have titles and abstracts on one lineL
    pubmed_input = re.sub(r"\n\s{6}", " ", pubmed_input)

    for line in pubmed_input.split("\n"):
        if re.match("PMID", line):
            PMID = re.search(r"\d+", line).group()
        if re.match("AB", line):
            spaetzle = re.findall(r"([^.]*?Spaetzle[^.]*\.)", line)
            if spaetzle:
                print("PubMedID: ", PMID, " ", spaetzle)
#&gt; PubMedID:  32591083   [' These events promote the ventral processing of Spaetzle, a ligand for Toll, which ultimately sets up the embryonic dorsal-ventral axis.']
#&gt; PubMedID:  31719046   ['  The dynamics indicate that a sharp extracellular gradient is formed through diffusion-based shuttling of the Spaetzle (Spz) morphogen that progresses through several nuclear divisions.']
#&gt; PubMedID:  27314646   [' While cytokines activating immune responses,  such as Spaetzle or Unpaired-3, have been identified and
```  
  
We will learn how to do this next week!

---
background-color: #f2f5eb

## PubMed search (cont.)

While this search was relatively trivial, you may like to do this kind of
search for a dozen or more genes, and possibly repeat the search periodically.

In that case, using Python to do the searches and parse the results can be a
huge time-saver!
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
