<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #20794d; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #20794d; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>pracs-sp21: Exercises: Week 4</title>


  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>


  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="pracs-sp21: Exercises: Week 4"/>
  <meta property="og:type" content="article"/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="pracs-sp21"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="pracs-sp21: Exercises: Week 4"/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","output"]}},"value":[{"type":"character","attributes":{},"value":["Exercises: Week 4"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","toc_depth"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]},{"type":"integer","attributes":{},"value":[3]}]}]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  </style>

  <script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
              .filter(function(item) { return !!item.description; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <style type="text/css">
  /* base variables */

  /* Edit the CSS properties in this file to create a custom
     Distill theme. Only edit values in the right column
     for each row; values shown are the CSS defaults.
     To return any property to the default,
     you may set its value to: unset
     All rows must end with a semi-colon.                      */

  /* Optional: embed custom fonts here with `@import`          */
  /* This must remain at the top of this file.                 */



  html {
    /*-- Main font sizes --*/
    --title-size:      50px;
    --body-size:       1.06rem;
    --code-size:       14px;
    --aside-size:      12px;
    --fig-cap-size:    13px;
    /*-- Main font colors --*/
    --title-color:     #000000;
    --header-color:    rgba(0, 0, 0, 0.8);
    --body-color:      rgba(0, 0, 0, 0.8);
    --aside-color:     rgba(0, 0, 0, 0.6);
    --fig-cap-color:   rgba(0, 0, 0, 0.6);
    /*-- Specify custom fonts ~~~ must be imported above   --*/
    --heading-font:    sans-serif;
    --mono-font:       monospace;
    --body-font:       sans-serif;
    --navbar-font:     sans-serif;  /* websites + blogs only */
  }

  /*-- ARTICLE METADATA --*/
  d-byline {
    --heading-size:    0.6rem;
    --heading-color:   rgba(0, 0, 0, 0.5);
    --body-size:       0.8rem;
    --body-color:      rgba(0, 0, 0, 0.8);
  }

  /*-- ARTICLE TABLE OF CONTENTS --*/
  .d-contents {
    --heading-size:    18px;
    --contents-size:   13px;
  }

  /*-- ARTICLE APPENDIX --*/
  d-appendix {
    --heading-size:    15px;
    --heading-color:   rgba(0, 0, 0, 0.65);
    --text-size:       0.8em;
    --text-color:      rgba(0, 0, 0, 0.5);
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  /* These properties only apply to Distill sites and blogs  */

  .distill-site-header {
    --title-size:       18px;
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  .distill-site-footer {
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  /*-- Additional custom styles --*/
  /* Add any additional CSS rules below                      */
  </style>
  <style type="text/css">
  /* base variables */

  /* Edit the CSS properties in this file to create a custom
     Distill theme. Only edit values in the right column
     for each row; values shown are the CSS defaults.
     To return any property to the default,
     you may set its value to: unset
     All rows must end with a semi-colon.                      */

  /* Optional: embed custom fonts here with `@import`          */
  /* This must remain at the top of this file.                 */



  html {
    /*-- Main font sizes --*/
    --title-size:      36px;
    --body-size:       1.06rem;
    --code-size:       14px;
    --aside-size:      12px;
    --fig-cap-size:    13px;
    /*-- Main font colors --*/
    --title-color:     #000000;
    --header-color:    rgba(0, 0, 0, 0.8);
    --body-color:      rgba(0, 0, 0, 0.8);
    --aside-color:     rgba(0, 0, 0, 0.6);
    --fig-cap-color:   rgba(0, 0, 0, 0.6);
    /*-- Specify custom fonts ~~~ must be imported above   --*/
    --heading-font:    sans-serif;
    --mono-font:       monospace;
    --body-font:       sans-serif;
    --navbar-font:     sans-serif;  /* websites + blogs only */
  }

  /*-- ARTICLE METADATA --*/
  d-byline {
    --heading-size:    0.6rem;
    --heading-color:   rgba(0, 0, 0, 0.5);
    --body-size:       0.8rem;
    --body-color:      rgba(0, 0, 0, 0.8);
  }

  /*-- ARTICLE TABLE OF CONTENTS --*/
  .d-contents {
    --heading-size:    18px;
    --contents-size:   13px;
  }

  /*-- ARTICLE APPENDIX --*/
  d-appendix {
    --heading-size:    15px;
    --heading-color:   rgba(0, 0, 0, 0.65);
    --text-size:       0.8em;
    --text-color:      rgba(0, 0, 0, 0.5);
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  /* These properties only apply to Distill sites and blogs  */

  .distill-site-header {
    --title-size:       18px;
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  .distill-site-footer {
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  /* --------------------------------------------------------------------------/*
  /* --------------------------------------------------------------------------/*
  /*-- Additional custom styles --*/
  /* Add any additional CSS rules below                      */

  code {
    background-color: #f0f0f0;
    border-radius: 5px;
  }

  pre code {
    background-color: white;
  }

  .sourceCode {
    background-color: #f0f0f0;
    /*padding: 10px 15px;*/
    /*border-left-width: 5px;
    border-left-style: solid;
    border-left-color: #b3dbff;*/
  }

  /*
  d-article pre {

  }*/

  d-article h3 {
    color: #bb0000;
  }

  /* --------------------------------------------------------------------------/*
  /* details sections */

  details > summary {
    color: black;
    padding: 2px 6px;
    background-color: #f0f8ff;
    border: none;
    box-shadow: 3px 3px 4px black;
    margin-bottom: 3px;
  }

  details > p:first-of-type {
    margin-top: -5px;
  }

  details {
    color: #707070;
    margin-left: 5px;
    padding: 10px 15px;
    /*background-color: #f0f8ff;
    border-left-width: 5px;
    border-left-style: solid;
    border-left-color: #b3dbff;
    border-radius: 0;*/
  }

  /******************
   * Coloured content boxes
   ****************/

  .content-box {
    box-sizing: border-box;
    background-color: #e2e2e2;
  }

  .content-box-q,
  .content-box-answer,
  .content-box-diy,
  .content-box-warning,
  .content-box-info {
    box-sizing: border-box;
    border-radius: 15px;
    margin: 0 0 15px;
    overflow: hidden;
    padding: 15px 20px 0px 70px;
    width: 100%;
    position: relative;
  }

  .content-box-info {
    background-color: #dfedff;
    border: solid 5px #c0dffa;
  }

  .content-box-warning {
    background-color: #ffcccc;
    border: solid 5px #ed8080;
  }

  .content-box-diy {
    background-color: #e4ede4;
    border: solid 5px #50ab50;
  }

  .content-box-q {
    background-color: #f4e9f5;
    border: solid 5px #ce84d1;
  }

  .content-box-answer {
    background-color: #e4ede4;
    border: solid 5px #50ab50;
  }

  .content-box-info:before,
  .content-box-warning:before,
  .content-box-diy:before,
  .content-box-q:before,
  .content-box-answer:before {
      font-family: FontAwesome;
      font-style: normal;
      font-weight: normal;
      text-decoration: inherit;
      color: #000;
      font-size: 25px;
      padding-right: 0.5em;
      position: absolute;
      top: 15px;
      left: 20px;
  }

  .content-box-info:before {
      content: "\f05a";
  }

  .content-box-warning:before {
      content: "\f071";
  }

  .content-box-diy:before {
      content: "\f044";
  }

  .content-box-q:before {
      content: "\f128";
  }

  .content-box-answer:before {
      content: "\f00c";
  }</style>
  <style type="text/css">
  /* base style */

  /* FONT FAMILIES */

  :root {
    --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  body,
  .posts-list .post-preview p,
  .posts-list .description p {
    font-family: var(--body-font), var(--body-default);
  }

  h1, h2, h3, h4, h5, h6,
  .posts-list .post-preview h2,
  .posts-list .description h2 {
    font-family: var(--heading-font), var(--heading-default);
  }

  d-article div.sourceCode code,
  d-article pre code {
  	font-family: var(--mono-font), var(--mono-default);
  }


  /*-- TITLE --*/
  d-title h1,
  .posts-list > h1 {
    color: var(--title-color, black);
  }

  d-title h1 {
    font-size: var(--title-size, 50px);
  }

  /*-- HEADERS --*/
  d-article h1,
  d-article h2,
  d-article h3,
  d-article h4,
  d-article h5,
  d-article h6 {
    color: var(--header-color, rgba(0, 0, 0, 0.8));
  }

  /*-- BODY --*/
  /* only text inside of <p> tags */
  d-article > p {
    color: var(--body-color, rgba(0, 0, 0, 0.8));
    font-size: var(--body-size, 1.06rem);
  }

  /*-- CODE --*/
  d-article div.sourceCode code,
  d-article pre code {
  	font-size: var(--code-size, 14px);
  }

  /*-- ASIDE --*/
  d-article aside {
  	font-size: var(--aside-size, 12px);
  	color: var(--aside-color, rgba(0, 0, 0, 0.6));
  }

  /*-- FIGURE CAPTIONS --*/
  figure .caption,
  figure figcaption,
  .figure .caption {
  	font-size: var(--fig-cap-size, 13px);
  	color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
  }

  /*-- METADATA --*/
  d-byline h3 {
  	font-size: var(--heading-size, 0.6rem);
  	color: var(--heading-color, rgba(0, 0, 0, 0.5));
  }

  d-byline {
  	font-size: var(--body-size, 0.8rem);
  	color: var(--body-color, rgba(0, 0, 0, 0.8));
  }

  d-byline a,
  d-article d-byline a {
  	color: var(--body-color, rgba(0, 0, 0, 0.8));
  }

  /*-- TABLE OF CONTENTS --*/
  .d-contents nav h3 {
  	font-size: var(--heading-size, 18px);
  }

  .d-contents nav a {
  	font-size: var(--contents-size, 13px);
  }

  /*-- APPENDIX --*/
  d-appendix h3 {
    font-size: var(--heading-size, 15px);
  	color: var(--heading-color, rgba(0, 0, 0, 0.65));
  }

  d-appendix {
    font-size: var(--text-size, 0.8em);
  	color: var(--text-color, rgba(0, 0, 0, 0.5));
  }

  d-appendix d-footnote-list a.footnote-backlink {
  	color: var(--text-color, rgba(0, 0, 0, 0.5));
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  .distill-site-header .title {
  	font-size: var(--title-size, 18px);
  	font-family: var(--navbar-font), var(--heading-default);
  }

  .distill-site-header a,
  .nav-dropdown .nav-dropbtn {
    font-family: var(--navbar-font), var(--heading-default);
  }

  .distill-site-header a:hover {
  	color: var(--hover-color, white);
  }

  .distill-site-header {
  	font-size: var(--text-size, 15px);
  	color: var(--text-color, rgba(255, 255, 255, 0.8));
  	background-color: var(--bkgd-color, #0F2E3D);
  }

  .distill-site-footer {
  	font-size: var(--text-size, 15px);
  	color: var(--text-color, rgba(255, 255, 255, 0.8));
  	background-color: var(--bkgd-color, #0F2E3D);
  }

  .distill-site-footer a:hover {
  	color: var(--hover-color, white);
  }</style>
  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.6.3/header-attrs.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Exercises: Week 4","authors":[]}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a href="index.html" class="title">pracs-sp21</a>
<a href="index.html">
<i class="fas fa-home" aria-hidden="true"></i>
</a>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Further resources
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="w08_python-resources.html">Python resources</a>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Topic overviews
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="overviews_shell.html">Shell</a>
<a href="overviews_shellscript.html">Shell scripting</a>
<a href="overviews_slurm.html">OSC: SLURM jobs and software</a>
<a href="overviews_markdown.html">Markdown</a>
<span class="nav-dropdown-header">Git</span>
<span class="nav-dropdown-header">Python</span>
</div>
</div>
<div class="nav-dropdown">
<button class="nav-dropbtn">
Final project
 
<span class="down-arrow">&#x25BE;</span>
</button>
<div class="nav-dropdown-content">
<a href="final_project.html">General info</a>
<a href="finalproject_proposal.html">Checkpoint I: Proposal</a>
<span class="nav-dropdown-header">Checkpoint II: Draft</span>
<span class="nav-dropdown-header">Checkpoint III: Presentation</span>
<span class="nav-dropdown-header">Checkpoint IV: Submission</span>
</div>
</div>
<input id="distill-search" class="nav-search hidden" type="text" placeholder="Search..."/>
</div>
<div class="nav-right">
<a href="about.html">About</a>
<a href="https://library.ohio-state.edu/record=b8624007~S7">
<i class="fas fa-book-open" aria-hidden="true"></i>
</a>
<a href="https://library.ohio-state.edu/record=b8538907~S7">
<i class="fas fa-book" aria-hidden="true"></i>
</a>
<a href="https://ondemand.osc.edu">
<i class="fas fa-desktop" aria-hidden="true"></i>
</a>
<a href="https://osu.instructure.com/courses/94975">
<i class="fas fa-large fa-graduation-cap" aria-hidden="true"></i>
</a>
<a href="https://github.com/mcic-osu/pracs-sp21" aria-label="Link to source">
<i class="fab fa-github" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Exercises: Week 4</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->

</div>


<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#main-exercises">Main exercises</a>
<ul>
<li><a href="#exercise-1-pseudogenes-with-grep-uniq--c-and-others">Exercise 1: Pseudogenes with <code>grep</code>, <code>uniq -c</code>, and others</a></li>
<li><a href="#exercise-2-calculations-with-awk">Exercise 2: Calculations with <code>awk</code></a></li>
<li><a href="#exercise-3-replacements-with-sed">Exercise 3: Replacements with <code>sed</code></a></li>
</ul></li>
<li><a href="#bonus-exercises">Bonus exercises</a>
<ul>
<li><a href="#exercise-4-fasta">Exercise 4: FASTA</a></li>
<li><a href="#exercise-5-exons">Exercise 5: Exons</a></li>
<li><a href="#exercise-6-miscellaneous">Exercise 6: Miscellaneous</a></li>
</ul></li>
<li><a href="#solutions">Solutions</a>
<ul>
<li><a href="#exercise-1-pseudogenes-with-grep-uniq--c-and-others-1">Exercise 1: Pseudogenes with <code>grep</code>, <code>uniq -c</code>, and others</a></li>
<li><a href="#exercise-2-calculations-with-awk-1">Exercise 2: Calculations with <code>awk</code></a></li>
<li><a href="#exercise-3-replacements-with-sed-1">Exercise 3: Replacements with <code>sed</code></a></li>
<li><a href="#exercise-4-fasta-1">Exercise 4: FASTA</a></li>
<li><a href="#exercise-5-exons-1">Exercise 5: Exons</a></li>
<li><a href="#exercise-6-miscellaneous-1">Exercise 6: Miscellaneous</a></li>
</ul></li>
</ul>
</nav>
</div>
<h2 id="main-exercises">Main exercises</h2>
<p>These exercises will use the same files that weve been working with in class: those in <code>chapter-07-unix-data-tools</code> in the <code>bds-files</code> repository for the Buffalo book.</p>
<p>You can either work directly in that directory, or create a sandbox directory with the Buffalo files copied.</p>
<p>For many of the exercises, there are multiple ways of solving them. The hints and solutions provide one or a few ways, but you may well end up using a different approach. If you struggle, try to follow the approach suggested by the hints, so you can later cross-check with solutions that use this same approach.</p>
<p>For most questions, the code required is quite similar to the examples in Buffalo chapter 7 (and the slides). If the hints are too cryptic for you, I recommend you first go back to the section on the relevant command in the chapter.</p>
<h3 id="exercise-1-pseudogenes-with-grep-uniq--c-and-others">Exercise 1: Pseudogenes with <code>grep</code>, <code>uniq -c</code>, and others</h3>
<p>In this exercise, we will start using the file <code>Mus_musculus.GRCm38.75_chr1.gtf</code> that weve seen in class.</p>
<p>Lets say that you need to extract annotation information for all pseudogenes on mouse chromosome 1 from the GTF, and write the results to a new GTF file. Well do that in this exercise with a detour along the way.</p>
<ol type="1">
<li><p>Count the number of lines with the string <em>pseudogene</em>.</p>
<details>
<p><summary>Hints</summary> Use <code>grep -c</code> to count the matches.</p>
</p>
</details></li>
</ol>
<p>Does this mean there are over a 1,000 pseudogenes on this chromosome? No, because there are nested genomic features in the file: for each gene, several subfeatures are listed, such as individual exons. This genomic feature type information is listed in column 3.</p>
<ol start="2" type="1">
<li><p>For lines that match pseudogene as above, which different feature types are listed in column 3, and what are their counts (i.e., create a count table for column 3)?</p>
<p>Are there any coding sequences (CDS), start codons (start_codon) and stop codons (stop_codon)?</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>To answer this question, compute count tables using <code>sort | uniq -c</code> after <code>grep</code>ping for pseudogene.</p>
</p>
</details></li>
<li><p>If you are more restrictive in your <code>grep</code> matching to only match pseudogene as a full word, do you still see CDS or start/stop codons?</p>
<details>
<p><summary>Hints</summary> Use <code>grep -w</code> to only match full words: consecutive alphanumeric characters and the underscore.</p>
<p>This would avoid matching longer words that <em>contain</em> pseudogene, such as transcribed_pseudogene.</p>
</p>
</details></li>
<li><p>We have exon, transcript and gene features in column 3, but we want just the <strong>gene-level coordinates</strong> of the pseudogenes: after <code>grep</code>ping for <em>pseudogene</em> like in the previous question, select only rows for which the feature type in column 3 is <code>gene</code>.</p>
<p>Pipe your output into <code>head</code> or <code>less</code>; we dont want to write to a new file yet.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>The gene you want to match is in its own column. Matching restrictively using the <code>-w</code> flag to only match whole words will work in this case.</p>
<p>Even better would be to <em>explicitly match the tabs</em> surrounding gene. This can be done using the <code>-P</code> flag to grep (<code>grep -P</code>) and the <code>\t</code> symbol for tabs.</p>
<p>Yet another approach that is even more explicit uses <code>awk</code> to make the match specifically for the 3rd column (recall that <code>$3</code> is the third column in <code>awk</code>, so you can use <code>$3 == "gene"</code>). If you try with <code>awk</code>, youll have to make sure to ignore the header lines, which are not tabular.</p>
</p>
</details></li>
<li><p>Comprehensively check that your output only contains gene in column 3.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Pipe your previous command into <code>cut</code> and then <code>sort | uniq -c</code> to quickly check which values are present in column 3.</p>
</p>
</details></li>
<li><p>Now, you are almost ready to write the results to a new GTF file, called <code>pseudogenes.gtf</code>, with the features you selected above: gene-level <em>pseudogene</em> features.</p>
<p>One challenge is that you also want the header, which you wont get when redirecting the output from the previous command.</p>
<p>A straightforward solution would be to first write the header to a new file, and then <em>append</em> the lines with the selected features.</p>
<p>Alternatively, you can use <em>subshells</em> to do this in a single go. If you want to try this, make sure you have read the Buffalo section on subshells (p.169-170) first.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Structure the subshell solution as follows:<br />
<code>(&lt;select-header&gt; ; &lt;select-from-rest-of-file&gt;) &gt; pseudogenes.gtf</code></p>
</p>
</details>
<p><br></p></li>
</ol>
<h3 id="exercise-2-calculations-with-awk">Exercise 2: Calculations with <code>awk</code></h3>
<p>In this exercise, you will continue working with the GTF file <code>Mus_musculus.GRCm38.75_chr1.gtf</code>.</p>
<ol type="1">
<li><p>How many elements are at least 1,000 base pairs long?</p>
<details>
<p><summary>Hints</summary></p>
<ul>
<li><p>Remove the header from consideration using <code>grep -v</code>.</p></li>
<li><p>Filter by feature length using <code>awk</code>, and get the length by subtracting column 4 from column 5.</p></li>
</ul>
</p>
</details></li>
<li><p>How many genes (gene in column 3) on the + strand (+ in column 7) are at least 1,000 base pairs long?</p>
<details>
<p><summary>Hints</summary></p>
<p>Adjust your command for the previous question as follows:</p>
<ul>
<li><p>Match columns 3 (feature type) and 7 (strand) with <code>awk</code>, using <code>==</code> to only select matching rows.</p></li>
<li><p>Use the <code>&amp;&amp;</code> operator in <code>awk</code> to combine conditions.</p></li>
</ul>
</p>
</details></li>
<li><p>Whats the combined size of all genes?<br />
(On this chromosome, that is: this GTF only contains a single chromosome.)</p>
<details>
<p><summary>Hints</summary></p>
<ul>
<li><p>First, remove the header from consideration using <code>grep -v</code>.</p></li>
<li><p>Next, select gene-level features only, which you can do with <code>awk</code> as above, with <code>grep -w</code>, or with <code>grep -P</code> with <code>\t</code> for tab.</p></li>
<li><p>Next, use a <strong>variable</strong> in <code>awk</code>:</p>
<ul>
<li>Initialize the variable in a <code>BEGIN</code> statement;</li>
<li>Add the length of the gene (column 5 - column 4) for each row;</li>
<li>Print the result with an <code>END</code> statement.</li>
</ul></li>
</ul>
</p>
</details></li>
<li><p>What is the mean UTR length?</p>
<details>
<p><summary>Hints</summary></p>
<p>The code is very similar to that for the previous question; just <code>grep</code> for UTR instead, and calculate the mean by dividing the final value of <code>awk</code> variable by the number of rows (<code>NR</code>) in the <code>END</code> statement.</p>
</p>
</details></li>
</ol>
<h3 id="exercise-3-replacements-with-sed">Exercise 3: Replacements with <code>sed</code></h3>
<p>The <code>genotypes.txt</code> contains diploid genotypes for 4 loci for 9 individuals.</p>
<ol type="1">
<li><p>Remove all underscores from <code>genotypes.txt</code> and write the output to a new file.</p>
<details>
<p><summary>Hints</summary></p>
<ul>
<li>Use the <code>sed 's/&lt;pattern&gt;/&lt;replacement&gt;/&lt;modifiers&gt;' &lt;file&gt;</code> construct.</li>
<li>Youll need the modifier to do <em>global</em> (i.e., &gt;1 per line) replacements.
</p></li>
</ul>
</details></li>
<li><p>Replace the <code>/</code> delimiter for genotypes (<code>G/G</code>) in <code>genotypes.txt</code> by a colon (<code>G:G</code>) and write the output to a new file.</p>
<details>
<p><summary>Hints</summary></p>
<ul>
<li>In the <code>sed 's/&lt;pattern&gt;/&lt;replacement&gt;/&lt;modifiers&gt;' &lt;file&gt;</code> construct, forward slashes <code>/</code> will not work here, because the pattern we are looking for is a forward slash. Use a different symbol instead, e.g.a <code>#</code>:<br />
<code>sed 's#&lt;pattern&gt;#&lt;replacement&gt;#&lt;modifiers&gt;' &lt;file&gt;</code>
</p></li>
</ul>
</details></li>
</ol>
<h2 id="bonus-exercises">Bonus exercises</h2>
<h3 id="exercise-4-fasta">Exercise 4: FASTA</h3>
<ol type="1">
<li>First, convert the FASTQ file <code>contaminated.fastq</code> to a FASTA file using the following <code>awk</code> command, and try to understand what it does, by and large:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">awk</span> <span class="st">&#39;{ if(NR%4 == 1) { printf(&quot;&gt;%s\n&quot;, substr($0,2)); } else if(NR%4 == 2) print; }&#39;</span> \</span>
<span id="cb1-2"><a href="#cb1-2"></a>    contaminated.fastq <span class="op">&gt;</span> contaminated.fasta</span></code></pre></div>
<ol start="2" type="1">
<li><p>Select FASTA entries (sequence <em>and</em> header) from <code>contaminated.fasta</code> with the sequence <code>AGATCGGAAGA</code> and write them to a new file, <code>contaminated_sel.fasta</code>. (Your new file should have 6 sequences, so 12 lines.)</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Match the sequence with <code>grep</code> and use the <code>-B</code> option to also retrieve the preceding line, which is the sequence header/identifier. In addition, use <code>--no-group-separator</code> to prevent grep from printing <code>--</code> between matches (alternatively, you could use another call to <code>grep</code> with <code>-v</code> to filter out <code>--</code> lines).</p>
</p>
</details></li>
<li><p>Check how many sequences, if any, have one ore more undetermined bases (<code>N</code>).</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Make sure you first exclude the header lines, since you only want to match <code>N</code> in the sequences.</p>
</p>
</details></li>
<li><p>Apparently, there may have been a problem with cycle 50 in the Illumina run.<br />
Get the frequency of each base for position 50 in the FASTA file.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Use <code>cut -c &lt;n&gt;</code> to extract the 50th position.</p>
</p>
</details></li>
</ol>
<p><br></p>
<h3 id="exercise-5-exons">Exercise 5: Exons</h3>
<p>In this exercise, you will once again work with the GTF file <code>Mus_musculus.GRCm38.75_chr1.gtf</code>.</p>
<p>Now, lets explore some statistics on <em>exons</em> in the GTF file.</p>
<ol type="1">
<li><p>Create a tab-separated, two-column file called <code>exons.txt</code> that lists each unique exon_id along with its gene_id, sorted by gene ID. (The exon_id and gene_id key-value pairs are in the huge final column.)</p>
<p>The first couple of lines of the file should look as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a>  <span class="ex">ENSMUSG00000000544</span>      ENSMUSE00000160546</span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="ex">ENSMUSG00000000544</span>      ENSMUSE00000160549</span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="ex">ENSMUSG00000000544</span>      ENSMUSE00000226264</span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="ex">ENSMUSG00000000544</span>      ENSMUSE00000395516</span></code></pre></div>
<details>
<p><summary>Hints</summary></p>
<p>
<ul>
<li><p>First <code>grep</code> for lines that contain gene_id followed by exon_id.</p></li>
<li><p>Use a <code>sed</code> command with two backreferences (capture the backreferences with parentheses, <code>(&lt;match&gt;)</code>, and recall them with <code>\1</code> and <code>\2</code>) to extract just the gene ID and the exon ID.</p></li>
<li><p>Finally, only keep unique lines and sort.</p></li>
</ul>
</p>
</details></li>
</ol>
<p>Use the <code>exons.txt</code> file you created to answer the following questions.</p>
<ol start="2" type="1">
<li><p>On average, how many exons does each gene have?</p>
<details>
<p><summary>Hints</summary></p>
<p>
<ul>
<li><p>The number of genes is simply the number of lines in <code>exons.txt</code>.</p></li>
<li><p>Count the number of exons using <code>uniq</code> and then <code>wc -l</code>.</p></li>
<li><p>The division can be done with <code>expr $n_genes / $n_exons</code> although you will not get any decimals (arithmetic in bash is very limited!). The above exmplae assumes you assigned the results to variables using command substitution, e.g.: <code>ngenes=$(wc -l ...)</code>.</p>
<p>(You could also, e.g., call Python inline, but were not there yet.)</p></li>
</ul>
</p>
</details></li>
<li><p>What is the highest number of exons for one gene?</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>Use <code>uniq -c</code> followed by a reverse numeric <code>sort</code>.</p>
</p>
</details></li>
<li><p>Create a count table for the number of exons per gene. It should tell you how many genes have one exon, how many have two exons, etc.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<ul>
<li><p>Youll need to process the <code>uniq -c</code> output of exon counts to get rid of the leading spaces. You can do this with <code>sed</code>; use <code>^ *</code> as the search pattern, or <code>^ +</code> with <code>sed -E</code> (the <code>+</code> symbol to match one ore more of the preceding character is in the <em>extended</em> regex set).</p></li>
<li><p>Then, youll need to create a count table from these counts, so another round of <code>count | uniq -c</code>.</p></li>
</ul>
</p>
</details></li>
</ol>
<h3 id="exercise-6-miscellaneous">Exercise 6: Miscellaneous</h3>
<p>In this exercise, you will once again work with the GTF file <code>Mus_musculus.GRCm38.75_chr1.gtf</code>.</p>
<ol type="1">
<li><p>Count the total number of features on each strand (<code>+</code> or <code>-</code>, column 7).</p></li>
<li><p>Count the number of features from each of the three values for gene_source (ensembl, havana, and ensembl_havana).</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>You can either use <code>sed</code> with a backreference to capture the gene_source type, or <code>grep -o</code> to capture a larger match followed by <code>cut</code> and <code>sed</code> to extract just the gene_source type.</p>
</p>
</details></li>
<li><p>How many genes have a UTR (untranslated region)? Create a list of gene_ids for genes with a UTR.</p>
<details>
<p><summary>Hints</summary></p>
<p>
<p>First <code>grep</code> for UTR, then for the gene_id while also extracting the match with <code>grep -o</code>. Once again, the extraction of a match can alternatively be done with a backreference in <code>sed</code>.</p>
</p>
</details></li>
</ol>
<h2 id="solutions">Solutions</h2>
<h3 id="exercise-1-pseudogenes-with-grep-uniq--c-and-others-1">Exercise 1: Pseudogenes with <code>grep</code>, <code>uniq -c</code>, and others</h3>
<p>To start with, I will assign the name of the GTF file to the variable <code>$gtf</code>, for quick reference in the solutions for the and the next exercises:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="va">gtf=</span>Mus_musculus.GRCm38.75_chr1.gtf</span></code></pre></div>
<details>
<summary>1. Count the number of lines with the string pseudogene.</summary>
<p>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a>$ <span class="fu">grep</span> -c <span class="st">&quot;pseudogene&quot;</span> <span class="va">$gtf</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt; 1041</span></span></code></pre></div>
The <code>-c</code> option to <code>grep</code> will count the number of matching lines.
</p>
</details>
<details>
<summary>2. Count feature types.</summary>
<p>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f 3 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">#&gt;     15 CDS</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">#&gt;    524 exon</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">#&gt;    223 gene</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">#&gt;      4 start_codon</span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co">#&gt;      4 stop_codon</span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">#&gt;    264 transcript</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">#&gt;      7 UTR</span></span></code></pre></div>
<p>The first line computes a count table for pseudogenes, and the second line computes an equivalent count table for all genes.</p>
<ul>
<li><p>We use <code>grep -v "^#"</code> to exclude the header (the command would work without this line, but its better not to assume the header wont match, and to explicitly remove it first). Note that the caret <code>^</code> is a regex symbol for the beginning of the line. We use the <code>-v</code> option to <code>grep</code> to invert the match.</p></li>
<li><p>We use <code>cut -f 3</code> to select the third column.</p></li>
<li><p>We use the <code>sort | uniq -c</code> combination to create a count table.</p></li>
</ul>
</p>
</details>
<details>
<summary>3. Count feature types after more restrictive matching.</summary>
<p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f 3 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt;    499 exon</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt;    221 gene</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt;    258 transcript</span></span></code></pre></div>
<p>This is very similar to the solution for the previous exercise, but now we use the <code>-w</code> option to <code>grep</code> to only select full words (consecutive alphanumeric characters and the underscore).</p>
</p>
</details>
<details>
<summary>4. Select <em>pseudogene</em> matches for which the feature type in column 3 is gene.</summary>
<p>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;gene&quot;</span> <span class="kw">|</span> <span class="fu">head</span></span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co"># Alternative 1: More explicit match by matching tabs:</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> -P <span class="st">&quot;\tgene\t&quot;</span> <span class="kw">|</span> <span class="fu">head</span></span>
<span id="cb7-5"><a href="#cb7-5"></a></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co"># Alternative 2: More explicit match by matching column 3 only:</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$3 == &quot;gene&quot;&#39;</span> <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<p>To match tabs like in the second solution, we need to use <code>grep</code>s <code>-P</code> option, for "Perl-like regular expressions. (Yes, it confusing and unfortunate that there are three types of regular expressions in <code>grep</code>!)</p>
<p>In the third solution we explicitly and exactly match column 3 with <code>awk</code>  this is the most robust solution.</p>
</p>
</details>
<details>
<summary>5. Check that your output only contains gene in column 3.</summary>
<p>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;pseudogene&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;gene&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>      <span class="fu">cut</span> -f 3 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb8-3"><a href="#cb8-3"></a></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; 223 gene</span></span></code></pre></div>
<p>We select the third column and use the <code>sort | uniq -c</code> combination to create a count table, which will quickly tell us whether there are multiple values in the column.</p>
</p>
</details>
<details>
<summary>6. Write the pseudogene selection to a new GTF file.</summary>
<p>
<p>Solution when writing the header and the rest of the file separately (dont forget to use <code>&gt;&gt;</code> for appending the output of the second line, if you use this method!):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a>$ <span class="fu">grep</span> <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="op">&gt;</span> pseudogenes.gtf</span>
<span id="cb9-2"><a href="#cb9-2"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>      <span class="fu">awk</span> <span class="st">&#39;$2 == &quot;pseudogene&quot; &amp;&amp; $3 == &quot;gene&quot;&#39;</span> \</span>
<span id="cb9-4"><a href="#cb9-4"></a>      <span class="op">&gt;&gt;</span> pseudogenes.gtf</span></code></pre></div>
<p>Solution with subshells:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a>$ <span class="kw">(</span><span class="fu">grep</span> <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span><span class="kw">;</span> <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>      <span class="fu">awk</span> <span class="st">&#39;$2 == &quot;pseudogene&quot; &amp;&amp; $3 == &quot;gene&quot;&#39;</span><span class="kw">)</span> <span class="kw">\</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>      <span class="op">&gt;</span> <span class="ex">pseudogenes.gtf</span></span></code></pre></div>
<p>Note: rather than with <code>grep -v "^#"</code>, we can also exclude the header using <code>!/^#/</code> in <code>awk</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a>$ <span class="kw">(</span><span class="fu">grep</span> <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span><span class="kw">;</span> <span class="kw">\</span></span>
<span id="cb11-2"><a href="#cb11-2"></a>      <span class="fu">awk</span> <span class="st">&#39;!/^#/ &amp;&amp; $2 == &quot;pseudogene&quot; &amp;&amp; $3 == &quot;gene&quot;&#39;</span> <span class="va">$gtf</span><span class="kw">)</span> <span class="kw">\</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>      <span class="op">&gt;</span> <span class="ex">pseudogenes.gtf</span></span></code></pre></div>
</p>
</details>
<h3 id="exercise-2-calculations-with-awk-1">Exercise 2: Calculations with <code>awk</code></h3>
<details>
<summary>1. How many features are at least 1,000 bp long?</summary>
<p>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>      <span class="fu">awk</span> <span class="st">&#39;$5 - $4 &gt; 999&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb12-3"><a href="#cb12-3"></a>      <span class="fu">wc</span> -l</span>
<span id="cb12-4"><a href="#cb12-4"></a></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">#&gt; 8773</span></span></code></pre></div>
<ul>
<li><p>First, we again exclude the header like in exercise 1 (this can also be done using <code>awk</code>: see below).</p></li>
<li><p>Next we subtract column 4 (<code>$4</code>) from column 5 (<code>$5</code>) and use as a condition that this difference should be larger than 999.</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># Alternative that excludes the header in awk:</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>$ <span class="fu">awk</span> <span class="st">&#39;!/^#/ &amp;&amp; $5 - $4 &gt; 999&#39;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>      <span class="fu">wc</span> -l</span>
<span id="cb13-4"><a href="#cb13-4"></a></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="co">#&gt; 8773</span></span></code></pre></div>
</p>
</details>
<details>
<summary>2. How many genes on the + strand are at least 1,000 bp long?</summary>
<p>
<div class="sourceCode" id="cb14"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>      <span class="fu">awk</span> <span class="st">&#39;$3 == &quot;gene&quot; &amp;&amp; $7 == &quot;+&quot; &amp;&amp; $5 - $4 &gt; 999&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>      <span class="fu">wc</span> -l</span>
<span id="cb14-4"><a href="#cb14-4"></a></span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">#&gt; 725</span></span></code></pre></div>
<ul>
<li><p>We chain together different conditions using <code>&amp;&amp;</code>.</p></li>
<li><p>The second line uses <code>awk</code> to match <code>+</code> in column 7, and to select lines where the difference between column 5 and 4 (i.e., the feature length) is larger than 999 bp.</p></li>
<li><p>The third line simply counts the number of lines we have left.</p></li>
</ul>
</p>
</details>
<details>
<summary>2. Whats the combined size of all genes?</summary>
<p>
<div class="sourceCode" id="cb15"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb15-2"><a href="#cb15-2"></a>      <span class="fu">grep</span> -w <span class="st">&quot;gene&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>      <span class="fu">awk</span> <span class="st">&#39;BEGIN { s=0 }; { s += $5 - $4 }; END { print s }&#39;</span></span>
<span id="cb15-4"><a href="#cb15-4"></a>  </span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="co">#&gt; 78935729    # 78,935,729 =&gt; ~79 Mbp</span></span></code></pre></div>
<ul>
<li><p>The first two lines are the same as for the previous question (see the solution there for further details).</p></li>
<li><p>In the third line, we use a variable in <code>awk</code> to compute a sum of the feature lengths (i.e., column 5 minus column 4). With the <code>BEGIN</code> statement, we initialize the variable <code>s</code> before we start processing lines. Then, for each line, we add the feature length to the value using the <code>+=</code> operator (shorthand for <code>s = s +</code>). Finally, with the <code>END</code> statement, we report the final value of the variable <code>s</code> after we have processed all lines.</p></li>
</ul>
</p>
</details>
<details>
<summary>3. What is the mean UTR length?</summary>
<p>
<div class="sourceCode" id="cb16"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>      <span class="fu">grep</span> -w <span class="st">&quot;UTR&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb16-3"><a href="#cb16-3"></a>      <span class="fu">awk</span> <span class="st">&#39;BEGIN { s=0 }; { s += $5 - $4 }; END { print s/NR }&#39;</span></span>
<span id="cb16-4"><a href="#cb16-4"></a>      </span>
<span id="cb16-5"><a href="#cb16-5"></a><span class="co">#&gt; 472.665      </span></span></code></pre></div>
<p>This code is very similar to that for the previous question (see that solution for more details), but now we divide the sum (variable <code>s</code>) by <code>NR</code>, an automatic variable in <code>awk</code> that holds the current record (line) number. Since the <code>END</code> statement is executed after all lines have been processed, the record number will be the <em>total</em> number of lines in the file.</p>
</p>
</details>
<h3 id="exercise-3-replacements-with-sed-1">Exercise 3: Replacements with <code>sed</code></h3>
<details>
<summary>1. In <code>genotypes.txt</code>, remove all underscores.</summary>
<p>
<div class="sourceCode" id="cb17"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1"></a>$ <span class="fu">sed</span> <span class="st">&#39;s/_//g&#39;</span> genotypes.txt <span class="op">&gt;</span> genotypes2.txt</span>
<span id="cb17-2"><a href="#cb17-2"></a></span>
<span id="cb17-3"><a href="#cb17-3"></a>$ <span class="fu">cat</span> genotypes2.txt</span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; id      indA    indB    indC    indD</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt; S000    G/G     A/G     A/A     A/G</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt; S001    A/T     A/T     T/T     T/T</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; S002    C/T     T/T     C/C     C/T</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt; S003    C/T     C/T     C/T     C/C</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">#&gt; S004    C/G     G/G     C/C     C/G</span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">#&gt; S005    A/T     A/T     A/T     T/T</span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">#&gt; S006    C/G     C/C     C/G     C/G</span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">#&gt; S007    A/G     G/G     A/A     G/G</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">#&gt; S008    G/T     G/T     T/T     G/T</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="co">#&gt; S009    C/C     C/C     A/A     A/C</span></span></code></pre></div>
Since there are multiple underscores on line 1, you need the <code>g</code> modifier.
</p>
</details>
<details>
<summary>2. In <code>genotypes.txt</code>, replace the <code>/</code> delimiter by a colon (<code>:</code>).</summary>
<p>
<div class="sourceCode" id="cb18"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1"></a>$ <span class="fu">sed</span> <span class="st">&#39;s#/#:#g&#39;</span> genotypes.txt <span class="op">&gt;</span> genotypes3.txt</span>
<span id="cb18-2"><a href="#cb18-2"></a></span>
<span id="cb18-3"><a href="#cb18-3"></a>$ <span class="fu">cat</span> genotypes3.txt</span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co">#&gt; id      ind_A   ind_B   ind_C   ind_D</span></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">#&gt; S_000   G:G     A:G     A:A     A:G</span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt; S_001   A:T     A:T     T:T     T:T</span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">#&gt; S_002   C:T     T:T     C:C     C:T</span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">#&gt; S_003   C:T     C:T     C:T     C:C</span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">#&gt; S_004   C:G     G:G     C:C     C:G</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">#&gt; S_005   A:T     A:T     A:T     T:T</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">#&gt; S_006   C:G     C:C     C:G     C:G</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="co">#&gt; S_007   A:G     G:G     A:A     G:G</span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co">#&gt; S_008   G:T     G:T     T:T     G:T</span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="co">#&gt; S_009   C:C     C:C     A:A     A:C</span></span></code></pre></div>
<p>Because the pattern we want to match is a <code>/</code>, we use a different symbol in the <code>sed</code> syntax: here, I chose a <code>#</code>.</p>
</p>
</details>
<h3 id="exercise-4-fasta-1">Exercise 4: FASTA</h3>
<details>
<summary>2. Write sequences with <code>AGATCGGAAGA</code> into a new file.</summary>
<p>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1"></a>$ <span class="fu">grep</span> -B 1 --no-group-separator <span class="st">&quot;AGATCGGAAGA&quot;</span> contaminated.fasta \</span>
<span id="cb19-2"><a href="#cb19-2"></a>      <span class="op">&gt;</span> contaminated_sel.fasta</span>
<span id="cb19-3"><a href="#cb19-3"></a></span>
<span id="cb19-4"><a href="#cb19-4"></a>$ <span class="fu">wc</span> -l <span class="op">&lt;</span> contaminated_sel.fasta</span>
<span id="cb19-5"><a href="#cb19-5"></a></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co">#&gt; 12</span></span>
<span id="cb19-7"><a href="#cb19-7"></a></span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co"># Alternative to `--no-group-separator`;</span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co"># remove the lines with `--` afterwards:</span></span>
<span id="cb19-10"><a href="#cb19-10"></a>$ <span class="fu">grep</span> -B 1 <span class="st">&quot;AGATCGGAAGA&quot;</span> contaminated.fasta <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb19-11"><a href="#cb19-11"></a>      <span class="fu">grep</span> -v <span class="st">&quot;--&quot;</span> <span class="op">&gt;</span> contaminated_sel.fasta</span></code></pre></div>
<ul>
<li><p>Use <code>B 1</code> to get the sequence ID line for each match.</p></li>
<li><p>Use <code>--no-group-separator</code> (top solution) or pipe into <code>grep -v "--"</code> to avoid having a group separator, which grep inserts for multiline output like this, in the output.</p></li>
</ul>
</p>
</details>
<details>
<summary>3. Check how many sequences have one ore more undetermined bases (N).</summary>
<p>
<p>Make sure you first exclude the header lines, since you only want to match N in the sequences themselves:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^&gt;&quot;</span> contaminated.fasta <span class="kw">|</span> <span class="fu">grep</span> -c <span class="st">&quot;N&quot;</span></span>
<span id="cb20-2"><a href="#cb20-2"></a></span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="co">#&gt; 0</span></span></code></pre></div>
<p>There are no <code>N</code>s in the sequence.</p>
</p>
</details>
<details>
<summary>4. Get the frequency of each base for position 50 in the fasta file.</summary>
<p>
<div class="sourceCode" id="cb21"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^&gt;&quot;</span> contaminated.fasta <span class="kw">|</span> <span class="fu">cut</span> -c 50 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c </span>
<span id="cb21-2"><a href="#cb21-2"></a></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="co">#&gt; 2 C</span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="co">#&gt; 3 G</span></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="co">#&gt; 3 T</span></span></code></pre></div>
</p>
</details>
<h3 id="exercise-5-exons-1">Exercise 5: Exons</h3>
<details>
<summary>1. Create a tab-delimited file with exon_id and gene_id</summary>
<p>
<div class="sourceCode" id="cb22"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1"></a>$ <span class="fu">grep</span> <span class="st">&quot;gene_id.*exon_id&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>      <span class="fu">sed</span> -n -E <span class="st">&#39;s/.*gene_id &quot;(\w+)&quot;.*exon_id &quot;(\w+).*/\1\t\2/p&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb22-3"><a href="#cb22-3"></a>      <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> exons.txt</span></code></pre></div>
<ul>
<li><p>In the first line of code, we select only lines that contain both gene_id and exon_id.</p></li>
<li><p>In the second line, we capture the values for gene_id and exon_id using backreferences in <code>sed</code>.</p></li>
<li><p>Finally, we sort (since we want to sort by the first column, just <code>sort</code> will work) and then take only unique rows using <code>uniq</code>, and redirect to a new file.</p></li>
</ul>
</p>
</details>
<details>
<summary>2. On average, how many exons does each gene have?</summary>
<p>
<div class="sourceCode" id="cb23"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1"></a>$ <span class="va">n_genes=$(</span><span class="fu">wc</span> -l <span class="op">&lt;</span> exons.txt<span class="va">)</span></span>
<span id="cb23-2"><a href="#cb23-2"></a>$ <span class="va">n_exons=$(</span><span class="fu">cut</span> -f 1 exons.txt <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> -l<span class="va">)</span></span>
<span id="cb23-3"><a href="#cb23-3"></a></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="co"># Divide using `expr`:</span></span>
<span id="cb23-5"><a href="#cb23-5"></a>$ <span class="fu">expr</span> <span class="va">$(</span><span class="fu">wc</span> -l <span class="op">&lt;</span> exons.txt<span class="va">)</span> / <span class="va">$(</span><span class="fu">cut</span> -f 1 exons.txt <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> -l<span class="va">)</span></span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co">#&gt; 11</span></span>
<span id="cb23-8"><a href="#cb23-8"></a></span>
<span id="cb23-9"><a href="#cb23-9"></a><span class="co"># Alternatively, do the division with Python:</span></span>
<span id="cb23-10"><a href="#cb23-10"></a>$ <span class="ex">python</span> -c <span class="st">&quot;print(</span><span class="va">$n_genes</span><span class="st"> / </span><span class="va">$n_exons</span><span class="st">)&quot;</span></span>
<span id="cb23-11"><a href="#cb23-11"></a></span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="co">#&gt; 11.627</span></span></code></pre></div>
<ul>
<li><p>First, we count the number of genes, then the number of exons, and we assign each value to a variable.</p></li>
<li><p>Then, we divide using <code>expr</code> (or Python to get decimals).</p></li>
</ul>
</p>
</details>
<details>
<summary>3. What is the highest number of exons for one gene?</summary>
<p>
<div class="sourceCode" id="cb24"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1"></a>$ <span class="fu">cut</span> -f 1 exons.txt <span class="kw">|</span> <span class="fu">uniq</span> -c <span class="kw">|</span> <span class="fu">sort</span> -rn <span class="kw">|</span> <span class="fu">head</span></span>
<span id="cb24-2"><a href="#cb24-2"></a></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="co">#&gt;    134 ENSMUSG00000026131</span></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="co">#&gt;    116 ENSMUSG00000066842</span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="co">#&gt;    112 ENSMUSG00000026207</span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="co">#&gt;    108 ENSMUSG00000026609</span></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="co">#&gt;    102 ENSMUSG00000006005</span></span>
<span id="cb24-8"><a href="#cb24-8"></a><span class="co">#&gt;     97 ENSMUSG00000037470</span></span>
<span id="cb24-9"><a href="#cb24-9"></a><span class="co">#&gt;     93 ENSMUSG00000026141</span></span>
<span id="cb24-10"><a href="#cb24-10"></a><span class="co">#&gt;     92 ENSMUSG00000026490</span></span>
<span id="cb24-11"><a href="#cb24-11"></a><span class="co">#&gt;     91 ENSMUSG00000048000</span></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="co">#&gt;     82 ENSMUSG00000073664</span></span></code></pre></div>
<p>ENSMUSG00000026131 has as many as 134 exons!!</p>
</p>
</details>
<details>
<summary>3. Create a count table for the number of exons per gene.</summary>
<p>
<div class="sourceCode" id="cb25"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1"></a>$ <span class="fu">cut</span> -f 1 exons.txt <span class="kw">|</span> <span class="fu">uniq</span> -c <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>      <span class="fu">sed</span> <span class="st">&#39;s/^ *//&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -d <span class="st">&quot; &quot;</span> -f 1 <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb25-3"><a href="#cb25-3"></a>      <span class="fu">sort</span> -rn <span class="kw">|</span> <span class="fu">uniq</span> -c <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb25-4"><a href="#cb25-4"></a>      <span class="fu">sort</span> -rn <span class="op">&gt;</span> exon_count_table.txt</span>
<span id="cb25-5"><a href="#cb25-5"></a></span>
<span id="cb25-6"><a href="#cb25-6"></a>$ <span class="fu">head</span> exon_count_table.txt     </span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="co">#&gt;    646 1   # 646 genes with 1 exon</span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="co">#&gt;    186 2   # 186 genes with 2 exons</span></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="co">#&gt;     94 3   # etc...</span></span>
<span id="cb25-10"><a href="#cb25-10"></a><span class="co">#&gt;     71 5</span></span>
<span id="cb25-11"><a href="#cb25-11"></a><span class="co">#&gt;     71 4</span></span>
<span id="cb25-12"><a href="#cb25-12"></a><span class="co">#&gt;     53 8</span></span>
<span id="cb25-13"><a href="#cb25-13"></a><span class="co">#&gt;     51 7</span></span>
<span id="cb25-14"><a href="#cb25-14"></a><span class="co">#&gt;     51 6</span></span>
<span id="cb25-15"><a href="#cb25-15"></a><span class="co">#&gt;     44 11</span></span>
<span id="cb25-16"><a href="#cb25-16"></a><span class="co">#&gt;     41 10</span></span></code></pre></div>
<ul>
<li><p>In the first line, we get the number of exons for each gene.</p></li>
<li><p>In the second line, we clean up the <code>uniq -c</code> output.</p></li>
<li><p>In the third line, we create the count table of the number of exons.</p></li>
<li><p>Finally we sort this and redirect the output to a file.</p></li>
</ul>
</p>
</details>
<h3 id="exercise-6-miscellaneous-1">Exercise 6: Miscellaneous</h3>
<details>
<summary>1. Count the number of elements on each strand.</summary>
<p>
<div class="sourceCode" id="cb26"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">cut</span> -f 7 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="co">#&gt; 40574 +</span></span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="co">#&gt; 40652 -</span></span></code></pre></div>
</p>
</details>
<details>
<summary>2. Count the number of elements for each of the three values for gene_source.</summary>
<p>
<div class="sourceCode" id="cb27"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1"></a><span class="co"># Capture the gene_source with sed:</span></span>
<span id="cb27-2"><a href="#cb27-2"></a>$ <span class="fu">sed</span> -n -E <span class="st">&#39;s/.*gene_source &quot;(\w+)&quot;.*/\1/p&#39;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb27-3"><a href="#cb27-3"></a>      <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb27-4"><a href="#cb27-4"></a></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="co">#&gt; 18209 ensembl</span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co">#&gt; 61089 ensembl_havana</span></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co">#&gt; 1928 havana</span></span>
<span id="cb27-8"><a href="#cb27-8"></a></span>
<span id="cb27-9"><a href="#cb27-9"></a><span class="co"># Alternatively, capture the gene_source with `grep -o`:</span></span>
<span id="cb27-10"><a href="#cb27-10"></a>$ <span class="fu">grep</span> -E -o <span class="st">&#39;gene_source &quot;\w+&quot;&#39;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb27-11"><a href="#cb27-11"></a>      <span class="fu">cut</span> -f2 -d<span class="st">&quot; &quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/&quot;//g&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb27-12"><a href="#cb27-12"></a>      <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span>
<span id="cb27-13"><a href="#cb27-13"></a>      </span>
<span id="cb27-14"><a href="#cb27-14"></a><span class="co">#&gt; 18209 ensembl</span></span>
<span id="cb27-15"><a href="#cb27-15"></a><span class="co">#&gt; 61089 ensembl_havana</span></span>
<span id="cb27-16"><a href="#cb27-16"></a><span class="co">#&gt; 1928 havana</span></span></code></pre></div>
</p>
</details>
<details>
<summary>3. How many genes have a UTR? Create a list of <code>gene_id</code>s for genes with a UTR.</summary>
<p>
<div class="sourceCode" id="cb28"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;UTR&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-2"><a href="#cb28-2"></a>   <span class="fu">grep</span> -E -o <span class="st">&#39;gene_id &quot;\w+&quot;&#39;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> -l</span>
<span id="cb28-3"><a href="#cb28-3"></a></span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="co">#&gt; 1179</span></span>
<span id="cb28-5"><a href="#cb28-5"></a></span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="co"># Create the list of genes:</span></span>
<span id="cb28-7"><a href="#cb28-7"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;UTR&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-8"><a href="#cb28-8"></a>      <span class="fu">grep</span> -E -o <span class="st">&#39;gene_id &quot;\w+&quot;&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-9"><a href="#cb28-9"></a>      <span class="fu">cut</span> -f2 -d<span class="st">&quot; &quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/&quot;//g&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-10"><a href="#cb28-10"></a>      <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> genes_with_UTR.txt</span>
<span id="cb28-11"><a href="#cb28-11"></a></span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co"># Alternative with sed:</span></span>
<span id="cb28-13"><a href="#cb28-13"></a>$ <span class="fu">grep</span> -v <span class="st">&quot;^#&quot;</span> <span class="va">$gtf</span> <span class="kw">|</span> <span class="fu">grep</span> -w <span class="st">&quot;UTR&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-14"><a href="#cb28-14"></a>      <span class="fu">sed</span> -n -E <span class="st">&#39;s/.*gene_id &quot;(\w+)&quot;.*/\1/p&#39;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb28-15"><a href="#cb28-15"></a>      <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="kw">|</span> <span class="fu">wc</span> -l <span class="op">&gt;</span> genes_with_UTR.txt</span></code></pre></div>
</p>
</details>
<div class="sourceCode" id="cb29"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<div class="distill-site-nav distill-site-footer">
<p>Jelmer Poelstra, 2020-2021</p>
</div>
<!--/radix_placeholder_navigation_after_body-->


</body>

</html>
